src/lib/ppx_coda/versioned_type.ml

```diff
@@ -82,6 +90,94 @@ let validate_module_version module_version loc =
          begin with 0, got: \"%s\""
         module_version
 
+[%%if
+print_versioned_types]
+
+let contains_deriving_bin_io (attrs : attributes) =
+  match
+    List.find attrs ~f:(fun ({txt; _}, _) -> String.equal txt "deriving")
+  with
+  | Some (_deriving, payload) -> (
+    match payload with
+    (* always have a tuple here; any deriving items are in addition to `version` *)
+    | PStr [{pstr_desc= Pstr_eval ({pexp_desc= Pexp_tuple items; _}, _); _}] ->
+        List.exists items ~f:(fun item ->
+            match item with
+            | {pexp_desc= Pexp_ident {txt= Lident "bin_io"; _}; _} ->
+                true
+            | _ ->
+                false )
+    | _ ->
+        false )
+  | None ->
+      (* unreachable *)
+      false
+
+(* singleton attribute *)
+let just_bin_io =
+  let loc = Location.none in
+  [ ( {txt= "deriving"; loc}
+    , PStr
+        [ { pstr_desc=
+              Pstr_eval
+                ( { pexp_desc= Pexp_ident {txt= Lident "bin_io"; loc}
+                  ; pexp_loc= loc
+                  ; pexp_attributes= [] }
+                , [] )
+          ; pstr_loc= loc } ] ) ]
+
+(* filter attributes from types, except for bin_io, don't care about changes to others *)
+let filter_type_decls_attrs type_decl =
+  (* retain `deriving bin_io` *)
+  let attrs = type_decl.ptype_attributes in
+  let ptype_attributes =
+    if contains_deriving_bin_io attrs then just_bin_io else []
+  in
+  {type_decl with ptype_attributes}
+
+(* convert type_decls to structure item so we can print it *)
+let type_decls_to_stri type_decls =
+  (* alas, type derivers aren't passed the rec flag, use a default *)
```

Comment:
`deriving` throws an error on non-recursive types. The flag should always be `Recursive` if we want to be consistent with the source code.